toffy/fov_watcher.py

```diff
@@ -0,0 +1,239 @@
+import os
+import time
+import json
+from datetime import datetime
+from typing import Callable, List, Tuple
+from watchdog.events import FileCreatedEvent, FileSystemEventHandler
+from watchdog.observers import Observer
+
+
+class RunStructure:
+    """Expected bin and json files
+
+    Attributes:
+        completion (dict): Whether or not an expected file has been created
+    """
+    def __init__(self, run_folder: str, timeout: int = 10 * 60):
+        """ initializes RunStructure by parsing run json within provided run folder
+
+        Args:
+            run_folder (str):
+                path to run folder
+            timeout (int):
+                number of seconds to wait for non-null filesize before raising an error
+        """
+        self.timeout = timeout
+        self.completion = {}
+
+        # find run .json and get parameters
+        with open(f'{run_folder}.json', 'r') as f:
+            run_metadata = json.load(f)
+
+        # parse run_metadata and populate expected structure
+        for fov in run_metadata.get('fovs', ()):
+            run_order = fov.get('runOrder', -1)
+            scan = fov.get('scanCount', -1)
+            if run_order * scan < 0:
+                raise KeyError(f"Could not locate keys in {run_folder}.json")
+
+            fov_name = f'fov-{run_order}-scan-{scan}'
+            self.completion[fov_name] = {
+                'json': False,
+                'bin': False,
+            }
+
+    def check_run_condition(self, path: str, check_interval: int = 10) -> Tuple[bool, str]:
+        """Checks if all requisite files exist and are complete
+
+        Args:
+            path (str):
+                path to expected file
+            check_interval (int):
+                number of seconds to wait before re-checking filesize
+
+        Raises:
+            TimeoutError
+
+        Returns:
+            (bool, str):
+                whether or not both json and bin files exist, as well as the name of the point
+        """
+
+        # TODO: check watchdog path depth
+
+        fov_name, extension = path.split('.')[0:2]
```

Comment:
this kind of ties into the `FileCreatedEvent` q so I'll answer both here.  Basically, the `on_created` function is called whenever **any** file is added to the watched directory/subdirectories.  So, each file that's copied from the instrument will call `on_created` which will directly pass that path to this function here, which will update the completion/progress dictionary if it has the relevant keys.

Reply:
that's my assumption. the only time this becomes an issue is in the case of something like `fov-1-scan-1.bin.evilfileextension` which I suppose could happen? I know Matlab likes doing this with *.m files; creating a copy of one called something like *.m._ or something? 

Anyways, I could validate that the length of `path.split('.')` is actually two, just in case.  That way it can throw a warning or something